core/src/p2p.rs

```diff
@@ -0,0 +1,571 @@
+                to_disconnect_keys.push(public_key.clone()); // Disconnect peers not in target.
+                continue;
+            }
+
+            let mut _byte = 0_u8;
+            if let Ok(byte_count) = stream.peek(std::slice::from_mut(&mut _byte)) { // Packet incomming
+                if byte_count == 0 {
+                    continue;
+                }
+                let mut packet_size = [0_u8; 4];
+                if stream.read_exact(&mut packet_size).is_err() // TODO: Peek so that we don't ever wait on packet completion.
+                {
+                    to_disconnect_keys.push(public_key.clone());
+                    continue;
+                }
+                let packet_size = u32::from_le_bytes(packet_size);
+                if packet_size == 0 { continue; }
+
+                let mut buf = vec![0_u8; packet_size as usize];
+                if stream.read_exact(&mut buf).is_err() {
+                    to_disconnect_keys.push(public_key.clone());
+                    continue;
+                }
+
+                if let Ok(data) = crypto.decrypt(buf) {
+                    let network_message_maybe = Decode::decode(&mut data.as_slice());
+                    if let Ok(network_message) = network_message_maybe {
+                        match network_message {
+                            SumeragiPacket(data) => {
+                                println!("RECV sumeragi");
+                                sumeragi.incoming_message(data.into_v1());
+                            }
+                            BlockSync(data) => {
+                                block_sync.incoming_message(*data);
+                            },
+                            Health => {}
+                        }
+                    } else {
+                        to_disconnect_keys.push(public_key.clone());
+                        continue;
+                    }
+                } else {
+                    to_disconnect_keys.push(public_key.clone());
+                    continue;
+                }
+            } else {
+                to_disconnect_keys.push(public_key.clone());
+                continue;
+            }
+        }
+
+        for key in to_disconnect_keys {
+            println!("Disconnected during read from {}.", key);
+            connected_to_peers.remove(&key).expect("Peer to disconnect must have been in the hashmap.");
+        }
+    }
+}
+
+pub fn start_listen_loop(p2p: Arc<P2PSystem>) -> ThreadHandler {
+    // Oneshot channel to allow forcefully stopping the thread.
+    let (shutdown_sender, shutdown_receiver) = tokio::sync::oneshot::channel();
+
+    let thread_handle = std::thread::spawn(move || {
+        p2p_listen_loop(&p2p, shutdown_receiver);
+    });
+
+    let shutdown = move || {
+        let _result = shutdown_sender.send(());
+    };
+
+    ThreadHandler::new(Box::new(shutdown), thread_handle)
+}
+
+// nocheckin replace println's with trace's.
+fn p2p_listen_loop(
```

Comment:
Prefixing the name of the function with its common namespace, and describing the function by its shape, rather than purpose, is an anti-pattern common in OOP. 

More importantly it is **not at all obvious** what's the difference between a `listen` and a `read` loop.